File: 10.py
# 캡스톤에 쓰이는 DC모터 방향조절(앞/뒤/좌/우), 거리 이동하기 

import RPi.GPIO as IO
import time
import math 

# DC 모터 왼쪽 (엔코더 O)
pwmPinA = 13 # 모터드라이버 ENA
AIN1 = 6 # IN 1
AIN2 = 5 # IN 2
encPinA = 2 # 보라색 (A)  
encPinB = 3 # 파랑색 (B)

# DC 모터 오른쪽 (엔코더 X)
pwmPinB = 21 # 모터드라이버 ENB
BIN3 = 4 # IN 3
BIN4 = 16 # IN 4
encPinC = 27 # 보라색 (C) - 20 
encPinD = 22 # 파랑색 (D) - 21

IO.setmode(IO.BCM)
IO.setwarnings(False)
IO.setup(encPinA, IO.IN, pull_up_down=IO.PUD_UP)
IO.setup(encPinB, IO.IN, pull_up_down=IO.PUD_UP)
IO.setup(encPinC, IO.IN, pull_up_down=IO.PUD_UP)
IO.setup(encPinD, IO.IN, pull_up_down=IO.PUD_UP)
IO.setup(pwmPinA, IO.OUT, initial=IO.LOW)
IO.setup(pwmPinB, IO.OUT, initial=IO.LOW)
IO.setup(AIN1, IO.OUT, initial=IO.LOW)
IO.setup(AIN2, IO.OUT, initial=IO.LOW)
IO.setup(BIN3, IO.OUT, initial=IO.LOW)
IO.setup(BIN4, IO.OUT, initial=IO.LOW)

p1 = IO.PWM(pwmPinA, 100)
p2 = IO.PWM(pwmPinB, 100)
p1.start(0)
p2.start(0)

encoderPosA = 0
encoderPosB = 0

# 인터럽트
def encoderA(encPinA):
    global encoderPosA
    if IO.input(encPinA) == IO.input(encPinB):
        encoderPosA += 1 
    else:
        encoderPosA -= 1
   
def encoderB(encPinB):
    global encoderPosA
    if IO.input(encPinA) == IO.input(encPinB):
        encoderPosA -= 1
    else:
        encoderPosA += 1

def encoderC(encPinC):
    global encoderPosB
    if IO.input(encPinC) == IO.input(encPinD):
        encoderPosB += 1 
    else:
        encoderPosB -= 1
   
def encoderD(encPinD):
    global encoderPosB
    if IO.input(encPinC) == IO.input(encPinD):
        encoderPosB -= 1
    else:
        encoderPosB += 1

IO.add_event_detect(encPinA, IO.BOTH, callback=encoderA)
IO.add_event_detect(encPinB, IO.BOTH, callback=encoderB)
IO.add_event_detect(encPinC, IO.BOTH, callback=encoderC)
IO.add_event_detect(encPinD, IO.BOTH, callback=encoderD)

# PID 제어
ratio = 360./144./40. # 한 바퀴에 약 13,728펄스 (정확하지 않음 - 계산값)

# PID 상수
kp = float(input("kp: "))
kd = 0.         
ki = 0.

# DC 모터 왼쪽
di_A = 0.
error_prev_A = 0.
error_prev_prev_A = 0.

# DC 모터 오른쪽
di_B = 0.
error_prev_B = 0.
error_prev_prev_B = 0.

dt = 0.
dt_sleep = 0.01
tolerance = 0.1

start_time = time.time()

time_prev = 0.
controlA = 0.
controlB = 0.

target_direction = 0. 
target_distance = 0.
rotation = 0.

r = 11.5    # 타이어의 반지름 

#----------------------------------- 모터의 이동 거리 -----------------------------------#
# 모터 이동 거리 (r: 11.5) 
try :
    target_direction = "front" 
    target_distance = int(input("이동하고 싶은 거리를 입력하시오: "))
    while True: 
        wheel = 2*math.pi*r     #원둘레 = 72.26
        target_deg = (360*target_distance / wheel) / 2    # 목표 각도
    
      # DC 모터 왼쪽
        motorDegA = encoderPosA * ratio     # 모터 움직인 각도
        errorA = target_deg - motorDegA     # 각도 오차값 
        de_A = errorA - error_prev_A
        di_A += errorA * dt
        dt = time.time() - time_prev
        
        delta_vA = kp*de_A + ki*errorA + kd*(errorA - 2*error_prev_A + error_prev_prev_A)
        controlA += delta_vA
        error_prev_A = errorA
        error_prev_prev_A = error_prev_A
    
        motor_distanceA = motorDegA * wheel / 360            # 모터 움직인 거리
        derrorA = abs(target_distance - motor_distanceA)     # 거리 오차값

        # DC 모터 오른쪽
        motorDegB = abs(encoderPosB * ratio)     # 모터 움직인 각도
        errorB = target_deg - motorDegB          # 각도 오차값 
        de_B = errorB - error_prev_B
        di_B += errorB * dt
        dt = time.time() - time_prev

        delta_vB = kp*de_B + ki*errorB + kd*(errorB - 2*error_prev_B + error_prev_prev_B)
        controlB += delta_vB
        error_prev_B = errorB
        error_prev_prev_B = error_prev_B

        motor_distanceB = motorDegB * wheel / 360           # 모터 움직인 거리
        derrorB = abs(target_distance - motor_distanceB)    # 거리 오차값

        # 전진 -------------------------------------------------
        if (target_direction == 'front') :  
            IO.output(AIN1, IO.LOW)
            IO.output(AIN2, IO.HIGH)
            IO.output(BIN3, IO.LOW)
            IO.output(BIN4, IO.HIGH)
            time.sleep(0.01)
            p1.ChangeDutyCycle(min(abs(controlA), 100))
            p2.ChangeDutyCycle(min(abs(controlB), 100))

            print('각도 = %5.1f' %(motorDegB))
            print('원하는 각도 = %5.1f' %(target_deg))
            print('ctrlA = %7.1f, degA = %5.1f, errA = %5.1f, disA = %5.1f, derrA = %5.1f' %(controlA, motorDegA, errorA, motor_distanceA, derrorA))  
            print('ctrlB = %7.1f, degB = %5.1f,s errB = %5.1f, disB = %5.1f, derrB = %5.1f' %(controlB, motorDegB, errorB, motor_distanceB, derrorB))  
            print('enc = %5.1f' %(encoderPosA))

            if ((motorDegA >= target_deg)):
                IO.output(AIN1, IO.LOW)
                IO.output(AIN2, IO.LOW) 
                IO.output(BIN3, IO.LOW)
                IO.output(BIN4, IO.LOW)

                time.sleep(0.01)
                p1.ChangeDutyCycle(0)
                p2.ChangeDutyCycle(0)

            time_prev = time.time()
            time.sleep(dt_sleep)

        # 후진 -------------------------------------------------------------------    
        elif (target_direction == 'back'): 
            IO.output(AIN1, IO.HIGH)
            IO.output(AIN2, IO.LOW)
            IO.output(BIN3, IO.HIGH)
            IO.output(BIN4, IO.LOW)
            time.sleep(0.01)
            p1.ChangeDutyCycle(min(abs(controlA), 100))
            p2.ChangeDutyCycle(min(abs(controlA), 100))
        
            print('ctrlA = %7.1f, degA = %5.1f, errA = %5.1f, disA = %5.1f, derrA = %5.1f' %(controlA, motorDegA, errorA, motor_distanceA, derrorA))  
            print('ctrlB = %7.1f, degB = %5.1f, errB = %5.1f, disB = %5.1f, derrB = %5.1f' %(controlB, motorDegB, errorB, motor_distanceB, derrorB)) 

            if ((motorDegA >= target_deg)):
                IO.output(AIN1, IO.LOW)
                IO.output(AIN2, IO.LOW) 
                IO.output(BIN3, IO.LOW)
                IO.output(BIN4, IO.LOW)

                time.sleep(0.01)
                p1.ChangeDutyCycle(0)
                p2.ChangeDutyCycle(0)

            time_prev = time.time()
            time.sleep(dt_sleep)
    
        # 오른쪽 -------------------------------------------------
        elif (target_direction == 'right') :  
            IO.output(AIN1, IO.LOW)
            IO.output(AIN2, IO.HIGH)
            IO.output(BIN3, IO.HIGH)
            IO.output(BIN4, IO.LOW)
            time.sleep(0.01)
            p1.ChangeDutyCycle(min(abs(controlA), 100))    
            p2.ChangeDutyCycle(min(abs(controlA), 100))  

            print('각도 = %5.1f' %(motorDegA))
            print('원하는 각도 = %5.1f' %(target_deg))
            print('ctrlA = %7.1f, degA = %5.1f, errA = %5.1f, disA = %5.1f, derrA = %5.1f' %(controlA, motorDegA, errorA, motor_distanceA, derrorA))  
            print('ctrlB = %7.1f, degB = %5.1f,s errB = %5.1f, disB = %5.1f, derrB = %5.1f' %(controlB, motorDegB, errorB, motor_distanceB, derrorB))  
            print('enc = %5.1f' %(encoderPosA))

            if ((motorDegA >= target_deg)):
                IO.output(AIN1, IO.LOW)
                IO.output(AIN2, IO.LOW) 
                IO.output(BIN3, IO.LOW)
                IO.output(BIN4, IO.LOW)

                time.sleep(0.01)
                p1.ChangeDutyCycle(0)
                p2.ChangeDutyCycle(0)

            time_prev = time.time()
            time.sleep(dt_sleep)

        # 왼쪽 -------------------------------------------------------------------    
        elif (target_direction == 'left'): 
            IO.output(AIN1, IO.LOW)  
            IO.output(AIN2, IO.HIGH)
            IO.output(BIN3, IO.LOW)
            IO.output(BIN4, IO.HIGH)
            time.sleep(0.01)
            p1.ChangeDutyCycle(min(abs(controlA), 30))
            p2.ChangeDutyCycle(min(abs(controlA), 100))
        
            print('ctrlA = %7.1f, degA = %5.1f, errA = %5.1f, disA = %5.1f, derrA = %5.1f' %(controlA, motorDegA, errorA, motor_distanceA, derrorA))  
            print('ctrlB = %7.1f, degB = %5.1f, errB = %5.1f, disB = %5.1f, derrB = %5.1f' %(controlB, motorDegB, errorB, motor_distanceB, derrorB)) 
    
            if ((motorDegA >= target_deg)):
                IO.output(AIN1, IO.LOW)
                IO.output(AIN2, IO.LOW) 
                IO.output(BIN3, IO.LOW)
                IO.output(BIN4, IO.LOW)

                time.sleep(0.01)
                p1.ChangeDutyCycle(0)
                p2.ChangeDutyCycle(0)

            time_prev = time.time()
            time.sleep(dt_sleep)


# Crtl + c 누르면 모터 작동 멈춤
except KeyboardInterrupt: 
    pass

p1.stop()
p2.stop() 

File: 10.cpp
/* PID 제어 */

#include "wiringPi.h"                   // analogRead(), pinMode(), delay() 함수 등 사용 
#include <softPwm.h>
#include <iostream>                    // C++ 입출력 라이브러리
#include <thread>
#include <chrono>
#include <ctime>
#include <string>
#include <unistd.h>
#include <algorithm>
#include <math.h>
#include <unistd.h>
#include <iomanip>

#define M_PI 3.14159265358979323846
using namespace std;

/* 핀 번호가 아니라 wiringPi 번호 ! */
/* gpio readall -> GPIO핀 / wiringPi핀 번호 확인법 */
/* ex) 핀 번호 8번, GPIO 14번, wiringPi 15번 */

// DC 모터 왼쪽 (엔코더 O)                                                      
#define pwmPinB 15              // 모터드라이버 ENA - GPIO핀 번호: 14 
#define BIN1 16                // IN1 - GPIO핀 번호: 15
#define BIN2 1                 // IN2 - GPIO핀 번호 : 18
#define encPinA 8               // 보라색 (A) - GPIO핀 번호 : 2
#define encPinB 9               // 파랑색 (B) - GPIO핀 번호 : 3

/* PID 제어 */
const float proportion = 360. / (84 * 10);       // 한 바퀴에 약 1350펄스 (정확하지 않음 - 계산값)

/* PID 상수 */
float kp = 0.; 
float kd = 0.;         
float ki = 0.;

float encoderPosRight;             // 엔코더 값 - 오른쪽

float motorDegB = 0;                   // 모터 각도B
float target_deg = 360.;

float errorB = 0; 
float error_prev_B = 0.;
float error_prev_prev_B = 0;

double controlB = 0.;
double tolerance = 0.1;

double de_B;
double di_B = 0;
double dt = 0;

double delta_vB;
double time_prev = 0;
double dt_sleep = 0.01;


std::time_t start_time = std::time(nullptr);

// 인터럽트 
void doEncoderA() {
  encoderPosRight  += (digitalRead(encPinA) == digitalRead(encPinB)) ? 1 : -1;
}
void doEncoderB() {
  encoderPosRight  += (digitalRead(encPinA) == digitalRead(encPinB)) ? -1 : 1;
}

int main(){
    wiringPiSetup();

    pinMode(encPinA, INPUT);
    pullUpDnControl(encPinA, PUD_UP);
    pinMode(encPinB, INPUT);
    pullUpDnControl(encPinB, PUD_UP);
    pinMode(pwmPinB, OUTPUT); 
    pinMode(BIN1, OUTPUT);
    pinMode(BIN2, OUTPUT);

    softPwmCreate(pwmPinB, 0, 255);
    softPwmWrite(pwmPinB, 0);

    digitalWrite(pwmPinB, LOW);
    digitalWrite(BIN1, LOW);
    digitalWrite(BIN2, LOW);

    wiringPiISR(encPinA, INT_EDGE_BOTH, &doEncoderA);
    wiringPiISR(encPinB, INT_EDGE_BOTH, &doEncoderB);

    // PID 상수 입력
    printf("kp: ");
    scanf("%f", &kp);
    printf("ki: ");
    scanf("%f", &ki);
    printf("kd: ");
    scanf("%f", &kd);

    while(true) {
        motorDegB = encoderPosRight * proportion;
        errorB = target_deg - motorDegB;
        de_B = errorB -error_prev_B;
        di_B += errorB * dt;
        dt = time(nullptr) - time_prev;
        
        delta_vB = kp*de_B + ki*errorB + kd*(errorB - 2*error_prev_B + error_prev_prev_B);
        controlB += delta_vB;
        error_prev_B = errorB;
        error_prev_prev_B = error_prev_B;

        int pwmB = std::min(std::abs(controlB), 100.); // 절댓값을 최대 100으로 맞춤
        analogWrite(pwmPinB, pwmB); // PWM 신호 출력
        digitalWrite(BIN1, controlB >= 0); // IN1 제어 신호 출력
        digitalWrite(BIN2, controlB <= 0); // IN2 제어 신호 출력
        
        std::cout << "time = " << std::setprecision(6) << std::fixed << std::time(nullptr) - start_time
          << ", enc = " << encoderPosRight
          << ", deg = " << motorDegB
          << ", err = " << errorB
          << ", ctrl = " << controlB << std::endl;

        if (std::abs(errorB) <= tolerance) {
            digitalWrite(BIN1, controlB >= 0);
            digitalWrite(BIN2, controlB <= 0);
            softPwmWrite(pwmPinB, 0);
            break;
        }

        time_prev = start_time;
        usleep(dt_sleep * 1000000);
    }
    
    return 0;
}